### PR TITLE
Include some debug output when the API server never appears.

### PR DESCRIPTION
**Pull request by mml**
Helps with #20916

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -744,6 +744,17 @@ function check-cluster() {
       local elapsed=$(($(date +%s) - ${start_time}))
       if [[ ${elapsed} -gt ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} ]]; then
           echo -e "${color_red}Cluster failed to initialize within ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} seconds.${color_norm}" >&2
+          if [[ $KUBE_TEST_DEBUG =~ ^[yY]$ ]]; then
```
Comment (ixdy) [inline cluster/gce/util.sh]: use `${KUBE_TEST_DEBUG-}"` so this doesn't error out if not set

File: cluster/gce/util.sh [hunk 2]
```diff
@@ -744,6 +744,17 @@ function check-cluster() {
...
+            local tmp_log=$(mktemp)
+            local file
+            for file in /var/log/startupscript.log /var/log/kube-apiserver.log; do
+              echo "$file contents:"
+              if gcloud compute copy-files ${MASTER_NAME}:$file $tmp_log; then
+                cat $tmp_log
+              fi
+            done
+            rm -f $tmp_log
```
Comment (ixdy) [inline cluster/gce/util.sh]: prefer `${}` variable syntax. also prefer putting in double quotes.

File: cluster/gce/util.sh [hunk 3]
```diff
@@ -744,6 +744,17 @@ function check-cluster() {
       local elapsed=$(($(date +%s) - ${start_time}))
       if [[ ${elapsed} -gt ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} ]]; then
           echo -e "${color_red}Cluster failed to initialize within ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} seconds.${color_norm}" >&2
+          if [[ $KUBE_TEST_DEBUG =~ ^[yY]$ ]]; then
+            local tmp_log=$(mktemp)
+            local file
+            for file in /var/log/startupscript.log /var/log/kube-apiserver.log; do
+              echo "$file contents:"
+              if gcloud compute copy-files ${MASTER_NAME}:$file $tmp_log; then
```
Comment (ixdy) [inline cluster/gce/util.sh]: include `--project "${PROJECT}" --zone "${ZONE}"`

File: cluster/gce/util.sh [hunk 4]
```diff
@@ -744,6 +744,17 @@ function check-cluster() {
       local elapsed=$(($(date +%s) - ${start_time}))
       if [[ ${elapsed} -gt ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} ]]; then
           echo -e "${color_red}Cluster failed to initialize within ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} seconds.${color_norm}" >&2
+          if [[ ${KUBE_TEST_DEBUG-} =~ ^[yY]$ ]]; then
+            local tmp_log="$(mktemp)"
+            local file
+            for file in /var/log/startupscript.log /var/log/kube-apiserver.log; do
+              echo "${file} contents:"
+              if gcloud compute copy-files "${MASTER_NAME}:${file}" "${tmp_log}"; then
+                cat "${tmp_log}"
```
Comment (ixdy) [inline cluster/gce/util.sh]: an alternative to `cat`ing these files: if `E2E_REPORT_DIR` is set (which it is on Jenkins runs), save these logs there; they are automatically uploaded to GCS.

File: cluster/gce/util.sh [hunk 5]
```diff
@@ -744,6 +744,17 @@ function check-cluster() {
       local elapsed=$(($(date +%s) - ${start_time}))
       if [[ ${elapsed} -gt ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} ]]; then
           echo -e "${color_red}Cluster failed to initialize within ${KUBE_CLUSTER_INITIALIZATION_TIMEOUT} seconds.${color_norm}" >&2
+          if [[ ${KUBE_TEST_DEBUG-} =~ ^[yY]$ ]]; then
+            local tmp_log="$(mktemp)"
+            local file
+            for file in /var/log/startupscript.log /var/log/kube-apiserver.log; do
+              echo "${file} contents:"
```
Comment (ixdy) [inline cluster/gce/util.sh]: may also be worth noting these are coming from the master, since it's not obvious otherwise?

**Comment by mml** [comment]
PTAL

**Comment by ixdy** [comment]
LGTM. For a full fix to #10423 one might want to save the logs to separate files instead of just dumping to stdout, but this is better than what we've got right now.

**Comment by ixdy** [comment]
@k8s-oncall may want to consider manually merging to help with #20916

**Comment by mml** [comment]
@ixdy thanks for the pointer.  I figured there must be something better than just catting the files.  Is there a magic directory we drop files into so that they get saved after each run?

**Comment by ixdy** [comment]
@mml yes - on Jenkins we set `E2E_REPORT_DIR` to a path that gets automatically uploaded.

**Pull request merged**